store/sqlstore/upgrade.go

```diff
@@ -1011,6 +1011,53 @@ func upgradeDatabaseToVersion535(sqlStore *SqlStore) {
 
 	sqlStore.CreateColumnIfNotExists("SidebarCategories", "Collapsed", "tinyint(1)", "boolean", "0")
 
+	sqlStore.CreateColumnIfNotExistsNoDefault("Channels", "TotalMsgCountRoot", "bigint", "bigint")
+	sqlStore.CreateColumnIfNotExistsNoDefault("Channels", "LastRootPostAt", "bigint", "bigint")
+	defer sqlStore.RemoveColumnIfExists("Channels", "LastRootPostAt")
+
+	// note: setting default 0 on pre-5.0 tables causes test-db-migration script to fail, so this column will be added to ignore list
+	sqlStore.CreateColumnIfNotExists("ChannelMembers", "MsgCountRoot", "bigint", "bigint", "0")
+	sqlStore.AlterColumnDefaultIfExists("ChannelMembers", "MsgCountRoot", model.NewString("0"), model.NewString("0"))
+
+	totalMsgCountRootCTE := `
+		SELECT Channels.Id channelid, COALESCE(COUNT(*),0) newcount, COALESCE(MAX(Posts.CreateAt), 0) as lastpost
+		FROM Channels
+		LEFT JOIN Posts ON Channels.Id = Posts.ChannelId
+		WHERE Posts.RootId = ''
+		GROUP BY Channels.Id
+	`
+	channelsCTE := "SELECT TotalMsgCountRoot, Id, LastRootPostAt from Channels"
+	updateChannels := `
+		WITH q AS (` + totalMsgCountRootCTE + `)
+		UPDATE Channels SET TotalMsgCountRoot = q.newcount, LastRootPostAt=q.lastpost
+		FROM q where q.channelid=Channels.Id;
+	`
+	updateChannelMembers := `
+		WITH q as (` + channelsCTE + `)
+		UPDATE ChannelMembers CM SET MsgCountRoot=TotalMsgCountRoot
+		FROM q WHERE q.id=CM.ChannelId AND LastViewedAt >= q.lastrootpostat;
+	`
+	if sqlStore.DriverName() == model.DATABASE_DRIVER_MYSQL {
+		updateChannels = `
+			UPDATE Channels
+			INNER Join (` + totalMsgCountRootCTE + `) as q
+			ON q.channelid=Channels.Id
+			SET TotalMsgCountRoot = q.newcount, LastRootPostAt=q.lastpost;
```

Comment:
Tried to run a comparison to test the perf impact of the PR but this query is still taking more than 1 hour to complete on MySQL. Interestingly the subquery (SELECT) only takes ~12 seconds.

Reply:
The main issue seems to be the JOIN on `Posts` which is not using an index and falls back to table-scanning:

```
*************************** 1. row ***************************
           id: 1
  select_type: UPDATE
        table: Channels
   partitions: NULL
         type: ALL
possible_keys: PRIMARY
          key: NULL
      key_len: NULL
          ref: NULL
         rows: 175270
     filtered: 100.00
        Extra: NULL
*************************** 2. row ***************************
           id: 1
  select_type: PRIMARY
        table: <derived2>
   partitions: NULL
         type: ref
possible_keys: <auto_key0>
          key: <auto_key0>
      key_len: 106
          ref: claudioltcomp1db.Channels.Id
         rows: 452345
     filtered: 100.00
        Extra: NULL
*************************** 3. row ***************************
           id: 2
  select_type: DERIVED
        table: Channels
   partitions: NULL
         type: index
possible_keys: PRIMARY,Name,idx_channels_team_id,idx_channels_name,idx_channels_update_at,idx_channels_create_at,idx_channels_delete_at,idx_channels_scheme_id,idx_channel_search_txt
          key: idx_channels_update_at
      key_len: 9
          ref: NULL
         rows: 175270
     filtered: 100.00
        Extra: Using index; Using temporary; Using filesort
*************************** 4. row ***************************
           id: 2
  select_type: DERIVED
        table: Posts
   partitions: NULL
         type: ALL
possible_keys: idx_posts_channel_id,idx_posts_root_id,idx_posts_channel_id_update_at,idx_posts_channel_id_delete_at_create_at
          key: NULL
      key_len: NULL
          ref: NULL
         rows: 2714070
     filtered: 16.67
        Extra: Using where; Using join buffer (Block Nested Loop)
4 rows in set (0.00 sec)
```

Reply:
Forcing the index seems to be helping:

#### Before:

```
mysql> UPDATE Channels
    -> INNER JOIN
    ->   (SELECT Channels.Id channelid,
    ->           COALESCE(COUNT(*), 0) newcount,
    ->           COALESCE(MAX(Posts.CreateAt), 0) AS lastpost
    ->    FROM Channels
    ->    LEFT JOIN Posts ON Channels.Id = Posts.ChannelId
    ->    WHERE Posts.RootId = ''
    ->    GROUP BY Channels.Id) AS q ON q.channelid=Channels.Id
    -> SET TotalMsgCountRoot = q.newcount,
    ->     LastRootPostAt=q.lastpost;
Query OK, 4 rows affected (5 hours 9 min 6.70 sec)
Rows matched: 4  Changed: 4  Warnings: 0
```

#### After

```
mysql> UPDATE Channels
    -> INNER JOIN
    ->   (SELECT Channels.Id channelid,
    ->           COALESCE(COUNT(*), 0) newcount,
    ->           COALESCE(MAX(Posts.CreateAt), 0) AS lastpost
    ->    FROM Channels
    ->    LEFT JOIN Posts
    ->    FORCE INDEX(idx_posts_channel_id) ON Channels.Id = Posts.ChannelId
    ->    WHERE Posts.RootId = ''
    ->    GROUP BY Channels.Id) AS q ON q.channelid=Channels.Id
    -> SET TotalMsgCountRoot = q.newcount,
    ->     LastRootPostAt=q.lastpost;
Query OK, 4 rows affected (12.66 sec)
Rows matched: 4  Changed: 4  Warnings: 0
```